docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,389 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
+
+```js
+print(json_obj.boom.dude.world);
+// RUNTIME ERROR: Uncaught TypeError: Cannot read properties of undefined (reading 'world')
+```
+
+Like in JavaScript, it is also possible to access object fields using `[]`:
+
+```js
+let foo = j["my-field"].your_field["their-field"];
+```
+
+To obtain an array of all the keys within a JSON object use the `Json.keys(o)` method. 
+
+
+```js
+let j = { hello: 123, world: [ 1, 2, 3 ] };
+assert(Json.keys(j) == ["hello", "world"]);
+```
+
+To obtain an array of all the values, use `Json.values(o)`. To obtain an array of all key/value
+pairs use `Json.entries(o)` (P2):
+
+```js
+assert(Json.values(j).equals(Json [ 123, [ 1, 2, 3 ] ]));
+assert(Json.entries(j).equals(Json [
+  [ "hello", 123 ],
+  [ "world", [ 1, 2, 3 ] ]
+]));
+```
+
+> NOTE: `values()` and `entries()` return an array inside a `Json` object because at the moment we
+> cannot represent heterogenous arrays in Wing.
+
+##### 1.1.4.3 JSON arrays
+
+To access an array element, use the `[]` notation:
+
+```js
+let item2: Json = json_array[2];
+```
+
+Trying to index a value that is not an array will return JavaScript `undefined`.
+
+##### 1.1.4.4 Assignment from native types
+
+It is also possible to assign the native `str`, `num`, `bool` and `Array<T>` values and they will
+implicitly be casted to `Json`:
+
+```js
+let my_str: str = "hello";
+let my_num: num = 183;
+let my_bool: bool = true;
+let my_arr: Array<num> = [1,2,3];
+
+let json_string = Json { 
+  a: my_string,
+  b: my_num,
+  c: my_bool,
+  d: my_arr
+};
+```
+
+
+##### 1.1.4.5 Assignment to native types
+
+We only allow implicit assignment from *safe* to *unsafe* types because otherwise we cannot
+guarantee safety (e.g. from `str` to `Json` but not from `Json` to `str`), so this won't work:
+
+```js
+let j = Json "hello";
+let s: str = j;
+//           ^ cannot assign `Json` to `str`.
+```
+
+To assign a `Json` to a strong-type variable, use the `from_json()` static method on the target
+type:
+
+```js
+let my_str = str.from_json(json_string);
+let my_number = num.from_json(json_number);
+let my_arr = Array<num>.from_json(json_array);
+```
+
+##### 1.1.4.6 Schema validation
+
+All `from_json()` methods will validate that the runtime type is compatible with the target type in
+order to ensure type safety (at a runtime cost):
+
+```js
+str.from_json(json_number);      // RUNTIME ERROR: unable to parse number `123` as a string.
+num.from_json(Json "\"hello\""); // RUNTIME ERROR: unable to parse string "hello" as a number
+
+let my_array = Json [1,2,3,"hello"];
+Array<num>.from_json(my_array); // RUNTIME ERROR: unable to parse `[1,2,3,"hello"]` as an array of `num`.
+```
+
+
+Use `unsafe: true` to disable this check at your own risk:
+
+```js
+let x = num.from_json(Json [1,2,3], unsafe: true);
+print(x + 5); // OUTPUT: 1,2,35
+```
+
+For each `from_json()`, there is a `try_from_json()` method which returns an optional `T?` which
```

Comment:
does `try_from_json()` have `unsafe: true` knob as well?

out of nowhere, was thinking `from_json(my_json, try: true)`, not sure if it fits winglang semantics overall

Reply:
> `from_json(my_json, try: true)`

That won't work because it implies that the return type of `from_json` must always be and optional `T?`.